analysis/taint/taint_utils_test.go

```diff
@@ -236,47 +239,174 @@ func expectTaintedCondInFuncs(funcNames ...string) func(error) bool {
+
+	result, err := taint.Analyze(cfg, program)
 	if err != nil {
-		for _, err := range result.State.CheckError() {
-			if !errorExpected(err) {
-				t.Errorf("taint analysis returned error: %v", err)
+		// t.Logf("taint analysis failed: %v", err) // use for debugging: sometimes errors are expected
+		if result.State != nil {
+			for _, err := range result.State.CheckError() {
+				if !errorExpected(err) {
+					t.Errorf("taint analysis returned error: %v", err)
+				}
 			}
 		}
 	}
 
-	expectSinkToSources, expectEscapeToSources := analysistest.GetExpectedTargetToSources(dir, ".")
-	checkExpectedPositions(t, program, result.TaintFlows, expectSinkToSources, expectEscapeToSources)
-	// Remove reports - comment if you want to inspect
-	os.RemoveAll(cfg.ReportsDir)
+	return runTestResult{
+		lp:  lp,
+		res: result,
+	}
 }
 
-func TestAll(t *testing.T) {
-	_, filename, _, _ := runtime.Caller(0)
-	dir := filepath.Join(filepath.Dir(filename), "..", "..", "testdata", "src", "taint", "basic")
-	err := os.Chdir(dir)
-	if err != nil {
-		panic(err)
+func setupConfig(cfg *config.Config, summarizeOnDemand bool) {
+	cfg.Options.ReportCoverage = false
+	cfg.Options.ReportPaths = false
+	cfg.Options.ReportSummaries = false
+	cfg.Options.ReportsDir = ""
+	cfg.LogLevel = int(config.ErrLevel) // change this as needed for debugging
+	cfg.SummarizeOnDemand = summarizeOnDemand
+}
+
+// sourceRegex matches an annotation of the form @Source(id1, id2 meta2, ...)
+// where the "argument" is either an identifier (e.g. id) or an identifier with
+// associated "metadata" (e.g. id call:example1->call:helper->call:example1$1).
+var sourceRegex = regexp.MustCompile(`//.*@Source\(((?:\s*(\w|(\w\s+[a-zA-Z0-9$:\->]+\s*))\s*,?)+)\)`)
+
+// sinkRegex matches annotations of the form "@Sink(id1, id2, id3)"
+var sinkRegex = regexp.MustCompile(`//.*@Sink\(((?:\s*\w\s*,?)+)\)`)
+
+// escapeRegex matches annotations of the form "@Escape(id1, id2, id3)"
+var escapeRegex = regexp.MustCompile(`//.*@Escape\(((?:\s*\w\s*,?)+)\)`)
+
+// expectedTaintTargetToSources analyzes the files in astFiles
+// and looks for comments @Source(id) and @Sink(id) to construct
+// expected flows from targets to sources in the form of two maps from:
+// - from sink positions to all the source position that reach that sink.
+// - from escape positions to the source of data that escapes.
+func expectedTaintTargetToSources(fset *token.FileSet, astFiles []*ast.File) (analysistest.TargetToSources, analysistest.TargetToSources) {
+	sink2source := make(analysistest.TargetToSources)
+	escape2source := make(analysistest.TargetToSources)
+	type sourceInfo struct {
+		meta string
+		pos  token.Position
 	}
-	sink2source, _ := analysistest.GetExpectedTargetToSources(dir, ".")
-	for sink, sources := range sink2source {
-		for source := range sources {
-			fmt.Printf("Source %s -> sink %s\n", formatutil.SanitizeRepr(source), formatutil.SanitizeRepr(sink))
+	sourceIDToSource := map[string]sourceInfo{}
```

Comment:
Was this code moved from analysistest to here because it's only used in taint tests now?

Reply:
Yes, it's taint-analysis-specific and `analysistest` seems like a more general package.